### ISSUE TITLE
Make user iconBackgrounds more flexible

### ISSUE THREAD

**Issue by djensen47** [NodeBB/NodeBB#6967]
According to this line of code:

https://github.com/NodeBB/NodeBB/blob/e17c4b2766d2080ee84e82564813a984fd776c38/src/user/data.js#L21

The background colors used for a the user icon background is hard-coded in an array. This can't be overridden even with styles. Instead of using inline styles what about using a class instead and then having default values set for that class?

```javascript
	var iconBackgrounds = [
		'icon-bg-01', 'icon-bg-02', 'icon-bg-03', 'icon-bg-04', 'icon-bg-05', 'icon-bg-06',
	];
```

Ans instead of setting `background-color` add this class instead.

Thoughts?

**Comment by barisusakli** [comment]
I have no objections to making the colors customizable 👍

**Comment by djensen47** [comment]
👍 Now that I have 2 changes, I'll submit a PR this week for this. 😁

I just need to get a dev environment setup locally.

**Comment by pitaj** [comment]
I recommend you don't make this a breaking change. That is, still provide the values for the inline style version so you don't need to fix every theme.

**Comment by djensen47** [comment]
> I recommend you don't make this a breaking change. That is, still provide the values for the inline style version so you don't need to fix every theme.

I don't think that will work. The problem is that inline styles always override classes.

What do you normally do for changes like this? Does it need to become a config or admin setting now?

**Comment by djensen47** [comment]
_Correction_: it looks like `!important` will actually override inline styles. (I should've known this)

I try not to use `!important` because it's usually feels like an anti-pattern, especially in this case. But if that's how we need to do it to be backward compatible, then that's the solution.

**Comment by pitaj** [comment]
> I try not to use `!important` because it's usually feels like an anti-pattern

You wouldn't need it. All you'd need to do is not include the inline style portion in themes which implement the class portion.

For instance, in Persona, you could replace this line:

https://github.com/NodeBB/nodebb-theme-persona/blob/226cb976cdd19d0500a8601d4a4a21d5d21d24c0/templates/flags/detail.tpl#L23

With this:

```html
<div class="media-object avatar avatar-lg {target.user.icon:bgColorClass}">{target.user.icon:text}</div>
```

Then there is no conflict, and you don't need `!important`

> Does it need to become a config or admin setting now?

If you're asking about how the colors should be configurable, I'd say that the class definitions should be up to the theme.

**Comment by julianlam** [comment]
While we're here, can someone take a look at feasibility of doing the text label with a pseudo-element instead? I have a sneaking suspicion that it might be:

* Easier to center vertically
* Play nicer with images

If we don't involve a text node inside... right now I think the text label avatars are a pixel higher than equivalent images, even though they are the same size in CSS.

**Comment by mySebbe** [comment]
The requested palette extensibility exists now through `filter:user.iconBackgrounds`; Core uses the returned list for avatar generation and the account picture chooser. That covers plugin-defined colors without changing Core. If the remaining request is specifically to replace inline colors with theme CSS classes/pseudo-elements, that is a separate theme/API migration and should be tracked independently. This issue can otherwise be closed as implemented.
